collectors/modality-probe-collector-common/src/lib.rs

```diff
@@ -283,320 +370,498 @@ impl TryFrom<&LogFileRow> for LogEntry {
+                        );
+                    } else {
+                        owned_report.event_log.push(EventLogEntry::Event(
+                            raw_entry
+                                .interpret_as_event_id()
+                                .ok_or_else(|| SerializationError::InvalidEventId(raw_entry))?,
+                        ));
+                    }
                 }
-                LogEvent::EventWithPayload(ev, payload) => {
-                    log_entries_buffer.push(LogEntry {
-                        session_id,
-                        segment_id: next_segment_id.into(),
-                        segment_index,
-                        probe_id,
-                        data: LogEntryData::EventWithPayload(*ev, *payload),
-                        receive_time,
-                    });
+                Next::Clock(id) => {
+                    let (epoch, ticks) = modality_probe::unpack_clock_word(raw);
+                    owned_report
+                        .event_log
+                        .push(EventLogEntry::TraceClock(LogicalClock { id, epoch, ticks }));
+                    interpret_next_as = Next::DontKnow;
+                }
+                Next::Payload(id) => {
+                    owned_report
+                        .event_log
+                        .push(EventLogEntry::EventWithPayload(id, raw));
+                    interpret_next_as = Next::DontKnow;
                 }
             }
-
-            segment_index += 1;
         }
-
-        next_segment_id += 1;
+        Ok(owned_report)
     }
+}
 
-    next_segment_id
+#[derive(Debug)]
+enum Next {
+    Clock(ProbeId),
+    Payload(EventId),
+    DontKnow,
 }
 
-pub fn add_owned_report_to_entries(
-    report: LogReport,
-    session_id: SessionId,
-    initial_segment_id: SegmentId,
-    receive_time: DateTime<Utc>,
-    log_entries_buffer: &mut Vec<LogEntry>,
-) -> u32 {
-    let mut next_segment_id = initial_segment_id.0;
-    let probe_id = report.probe_id;
-    for segment in report.segments {
-        let mut segment_index = 0;
-
-        for clock_bucket in segment.clocks {
-            log_entries_buffer.push(LogEntry {
-                session_id,
-                segment_id: next_segment_id.into(),
-                segment_index,
-                probe_id,
-                data: LogEntryData::LogicalClock(
-                    clock_bucket.id,
-                    clock_bucket.epoch,
-                    clock_bucket.ticks,
-                ),
-                receive_time,
-            });
+impl Report {
+    pub fn write_into_le_bytes(&self, bytes: &mut [u8]) -> Result<usize, SerializationError> {
+        if self.frontier_clocks.len() > std::u16::MAX as usize {
+            return Err(SerializationError::TooManyFrontierClocks(
+                self.frontier_clocks.len(),
+            ));
+        } else if self.event_log.len() > std::u32::MAX as usize {
```

Comment:
IIRC, the value used later in `set_n_log_entries` may actually be larger than this since more than one log entry may be used for, say, events with payloads. Thus a simple len check here is not sufficient.  Probably should do the check against the computed `num_u32_entries` below.

Reply:
Ah yeah, good catch
507ba4c9ab2b35a409247c8ced1e51303cbe5ab0